File: src/scheduler.cpp
#include "ai_usage/scheduler.h"

#include <algorithm>
#include <exception>
#include <stdexcept>

namespace ai_usage {

RefreshScheduler::RefreshScheduler(SnapshotCallback callback, std::unique_ptr<IClock> clock)
    : callback_(std::move(callback)), clock_(std::move(clock)) {
  if (!clock_) clock_ = CreatePlatformClock();
}

RefreshScheduler::~RefreshScheduler() { Stop(); }

void RefreshScheduler::SetProviders(std::vector<std::unique_ptr<IUsageProvider>> providers) {
  std::lock_guard lock(mutex_);
  if (started_) throw std::logic_error("cannot replace providers while scheduler is running");
  providers_.clear();
  for (auto& provider : providers) {
    const auto id = provider->Config().id;
    providers_.emplace(id, State{std::move(provider), TimePoint{}, 0U, true, false});
  }
}

void RefreshScheduler::SetInterval(std::chrono::minutes interval) {
  if (interval < std::chrono::minutes{1} || interval > std::chrono::minutes{60}) {
    throw std::invalid_argument("refresh interval must be 1..60 minutes");
  }
  std::lock_guard lock(mutex_);
  interval_ = interval;
  condition_.notify_all();
}

void RefreshScheduler::Start() {
  std::lock_guard lock(mutex_);
  if (started_) return;
  stopping_ = false;
  started_ = true;
  workers_.clear();
  workers_.emplace_back([this] { RunWorker(); });
  workers_.emplace_back([this] { RunWorker(); });
  schedulerThread_ = std::jthread([this] { Run(); });
}

void RefreshScheduler::Stop() {
  {
    std::lock_guard lock(mutex_);
    if (!started_) return;
    stopping_ = true;
    for (auto& [id, state] : providers_) {
      (void)id;
      if (state.running) state.provider->Cancel();
    }
    condition_.notify_all();
  }
  if (schedulerThread_.joinable()) schedulerThread_.join();
  for (auto& worker : workers_) {
    if (worker.joinable()) worker.join();
  }
  std::lock_guard lock(mutex_);
  workers_.clear();
  jobs_.clear();
  active_ = 0;
  for (auto& [id, state] : providers_) {
    (void)id;
    state.running = false;
  }
  started_ = false;
}

void RefreshScheduler::RefreshAll() {
  std::lock_guard lock(mutex_);
  for (auto& [id, state] : providers_) {
    (void)id;
    state.queued = true;
  }
  condition_.notify_all();
}

void RefreshScheduler::RefreshOne(const std::string& providerId) {
  std::lock_guard lock(mutex_);
  const auto found = providers_.find(providerId);
  if (found != providers_.end()) found->second.queued = true;
  condition_.notify_all();
}

void RefreshScheduler::NotifyClockChanged() {
  std::lock_guard lock(mutex_);
  condition_.notify_all();
}

std::chrono::seconds RefreshScheduler::Backoff(unsigned failures, const std::string& providerId) const {
  const unsigned exponent = failures == 0U ? 0U : std::min(failures - 1U, 7U);
  const auto base = std::chrono::seconds{30LL * (1LL << exponent)};
  const auto capped = std::min(base, std::chrono::duration_cast<std::chrono::seconds>(std::chrono::minutes{60}));
  const auto jitterRange = std::max<std::int64_t>(1, capped.count() / 5);
  const auto jitter = static_cast<std::int64_t>(std::hash<std::string>{}(providerId) % static_cast<std::size_t>(jitterRange));
  return std::min(capped + std::chrono::seconds{jitter},
                  std::chrono::duration_cast<std::chrono::seconds>(std::chrono::minutes{60}));
}

void RefreshScheduler::Refresh(State& state, TimePoint now) {
  ProviderSnapshot snapshot;
  try {
    snapshot = state.provider->Refresh();
    const auto validation = ValidateSnapshot(snapshot);
    if (!validation.valid) throw std::runtime_error(validation.errors.front());
  } catch (const ProviderException& exception) {
    snapshot.providerId = state.provider->Config().id;
    snapshot.displayName = state.provider->Config().name;
    snapshot.kind = state.provider->Config().kind;
    snapshot.observedAt = now;
    snapshot.freshness = Freshness::NoData;
    snapshot.health = Health::Error;
    snapshot.error = exception.Error();
  } catch (const std::exception& error) {
    snapshot.providerId = state.provider->Config().id;
    snapshot.displayName = state.provider->Config().name;
    snapshot.kind = state.provider->Config().kind;
    snapshot.observedAt = now;
    snapshot.freshness = Freshness::NoData;
    snapshot.health = Health::Error;
    snapshot.error = ProviderError{"refresh-failed", error.what(), true, std::nullopt};
  }

  {
    std::lock_guard lock(mutex_);
    if (snapshot.health == Health::Error) {
      ++state.failures;
      auto delay = Backoff(state.failures, state.provider->Config().id);
      if (snapshot.error.has_value() && snapshot.error->retryAfter.has_value()) delay = std::max(delay, *snapshot.error->retryAfter);
      state.nextDue = now + delay;
    } else {
      state.failures = 0;
      state.nextDue = now + interval_;
    }
  }
  callback_(snapshot);
}

void RefreshScheduler::Run() {
  while (!stopping_) {
    TimePoint nextWake = TimePoint::max();
    {
      std::unique_lock lock(mutex_);
      const auto now = clock_->Now();
      while (active_ < 2U) {
        auto selected = providers_.end();
        for (auto candidate = providers_.begin(); candidate != providers_.end(); ++candidate) {
          if (!candidate->second.running && (candidate->second.queued || candidate->second.nextDue <= now)) {
            selected = candidate;
            break;
          }
        }
        if (selected == providers_.end()) break;
        selected->second.queued = false;
        selected->second.running = true;
        jobs_.push_back(selected->first);
        ++active_;
        condition_.notify_all();
      }
      for (const auto& [id, state] : providers_) {
        (void)id;
        if (!state.running) {
          if (state.queued && active_ < 2U) nextWake = now;
          else nextWake = std::min(nextWake, state.nextDue);
        }
      }
      if (nextWake == TimePoint::max()) {
        condition_.wait(lock);
      } else {
        condition_.wait_until(lock, nextWake);
      }
    }
  }
}

void RefreshScheduler::RunWorker() {
  while (true) {
    std::string selected;
    {
      std::unique_lock lock(mutex_);
      condition_.wait(lock, [this] { return stopping_ || !jobs_.empty(); });
      if (stopping_) return;
      selected = std::move(jobs_.front());
      jobs_.pop_front();
    }
    State* state = nullptr;
    {
      std::lock_guard lock(mutex_);
      const auto found = providers_.find(selected);
      if (found != providers_.end()) state = &found->second;
    }
    if (state != nullptr && !stopping_) Refresh(*state, clock_->Now());
    {
      std::lock_guard lock(mutex_);
      const auto found = providers_.find(selected);
      if (found != providers_.end()) found->second.running = false;
      if (active_ > 0U) --active_;
      condition_.notify_all();
    }
  }
}

}  // namespace ai_usage

File: flutter_app/lib/scheduler/scheduler.dart
/// Refresh scheduler ported from the C++ reference: interval validation, at
/// most two concurrent refreshes, queued coalescing, the reference exponential
/// backoff with deterministic per-provider jitter, `Retry-After` override,
/// failure counting and error mapping to snapshots.
library;

import 'dart:async';

import '../domain/model.dart';
import '../domain/validate.dart';
import '../platform/interfaces.dart';
import '../providers/providers.dart';

/// Per-provider runtime state.
class ProviderState {
  ProviderState({required this.provider});

  final UsageProvider provider;
  DateTime nextDue = DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
  int failures = 0;
  bool queued = false;
  bool running = false;
}

/// Schedules provider refreshes on the shared interval with the reference's
/// backoff, honoring manual refreshes and a two-slot concurrency cap.
class RefreshScheduler {
  RefreshScheduler({
    required Clock clock,
    required this.onSnapshot,
    int maxConcurrent = 2,
  })  : _clock = clock,
        _maxConcurrent = maxConcurrent;

  final Clock _clock;
  final void Function(ProviderSnapshot snapshot) onSnapshot;
  final int _maxConcurrent;

  final Map<String, ProviderState> _providers = {};
  Duration _interval = const Duration(minutes: 5);
  Timer? _timer;
  int _active = 0;
  bool _started = false;
  final List<String> _jobs = [];

  /// Replaces the provider set; requires the scheduler to be stopped.
  void setProviders(List<UsageProvider> providers) {
    if (_started) {
      throw StateError('cannot replace providers while scheduler is running');
    }
    _providers.clear();
    for (final provider in providers) {
      _providers[provider.config.id] = ProviderState(provider: provider);
    }
  }

  /// Validates and applies the refresh interval (1..60 minutes).
  void setInterval(Duration interval) {
    if (interval < const Duration(minutes: 1) || interval > const Duration(minutes: 60)) {
      throw ArgumentError.value(interval, 'interval', 'refresh interval must be 1..60 minutes');
    }
    _interval = interval;
  }

  Duration get interval => _interval;

  void start() {
    if (_started) return;
    _started = true;
    _tick();
  }

  void stop() {
    _started = false;
    _timer?.cancel();
    _timer = null;
    _jobs.clear();
    _active = 0;
    _providers.forEach((_, state) => state.running = false);
  }

  /// Queues every provider for an immediate refresh; repeated calls while a
  /// refresh is pending coalesce into the same queued flag.
  void refreshAll() {
    _providers.forEach((_, state) {
      state.queued = true;
    });
    _wake();
  }

  /// Queues one provider by id.
  void refreshOne(String providerId) {
    final state = _providers[providerId];
    if (state != null) {
      state.queued = true;
    }
    _wake();
  }

  /// Notifies the scheduler that the clock changed (wake early).
  void notifyClockChanged() => _wake();

  void _wake() {
    if (!_started) return;
    _tick();
  }

  void _scheduleNext() {
    if (!_started) return;
    _timer?.cancel();
    _timer = null;
    _tick();
  }

  void _tick() {
    final now = _clock.now();
    // Start jobs while concurrency allows, in configuration order.
    for (final state in _providers.values) {
      if (_active >= _maxConcurrent) break;
      if (state.running) continue;
      if (state.queued || !state.nextDue.isAfter(now)) {
        state.queued = false;
        state.running = true;
        _active++;
        unawaited(_run(state, now));
      }
    }
    _timer ??= Timer(_nextWake(now), _scheduleNext);
  }

  Duration _nextWake(DateTime now) {
    var earliest = const Duration(days: 3650);
    for (final state in _providers.values) {
      if (state.running) continue;
      if (state.queued) return Duration.zero;
      final due = state.nextDue.difference(now);
      if (due < earliest) earliest = due;
    }
    return earliest < Duration.zero ? Duration.zero : earliest;
  }

  Future<void> _run(ProviderState state, DateTime now) async {
    final snapshot = await _refresh(state, now);
    // Update scheduling state under the same rules as the reference.
    final updatedNow = _clock.now();
    if (snapshot.health == Health.error) {
      state.failures++;
      var delay = backoff(state.failures, state.provider.config.id);
      final retryAfter = snapshot.error?.retryAfter;
      if (retryAfter != null && retryAfter > delay) {
        delay = retryAfter;
      }
      state.nextDue = updatedNow.add(delay);
    } else {
      state.failures = 0;
      state.nextDue = updatedNow.add(_interval);
    }
    state.running = false;
    _active = _active > 0 ? _active - 1 : 0;
    onSnapshot(snapshot);
    _scheduleNext();
  }

  /// Runs one refresh and maps failures onto error snapshots, mirroring the
  /// reference's `RefreshScheduler::Refresh`.
  Future<ProviderSnapshot> _refresh(ProviderState state, DateTime now) async {
    ProviderSnapshot snapshot;
    try {
      final result = await state.provider.refresh().timeout(
            _interval + const Duration(minutes: 1),
          );
      final validation = validateSnapshot(result);
      if (!validation.valid) {
        throw Exception(validation.errors.first);
      }
      snapshot = result;
    } catch (error) {
      snapshot = ProviderSnapshot(
        providerId: state.provider.config.id,
        displayName: state.provider.config.name,
        kind: state.provider.config.kind,
        observedAt: now,
        freshness: Freshness.noData,
        health: Health.error,
      );
      if (error is ProviderException) {
        snapshot.error = error.error;
      } else {
        snapshot.error = ProviderError('refresh-failed', error.toString(), transient: true);
      }
    }
    return snapshot;
  }

  /// The reference backoff: 30s * 2^(failures-1) capped at 60 minutes plus a
  /// deterministic per-provider jitter of up to one fifth of the capped base,
  /// with the sum capped at 60 minutes.
  Duration backoff(int failures, String providerId) {
    final exponent = failures == 0 ? 0 : (failures - 1).clamp(0, 7);
    final baseSeconds = 30 * (1 << exponent);
    final capped = baseSeconds > 3600 ? 3600 : baseSeconds;
    final jitterRange = capped ~/ 5 < 1 ? 1 : capped ~/ 5;
    final jitter = providerId.hashCode.abs() % jitterRange;
    final total = capped + jitter > 3600 ? 3600 : capped + jitter;
    return Duration(seconds: total);
  }

  /// Test/inspection hooks.
  bool isRunning(String providerId) => _providers[providerId]?.running ?? false;

  bool isQueued(String providerId) => _providers[providerId]?.queued ?? false;

  DateTime? nextDueFor(String providerId) => _providers[providerId]?.nextDue;
}